### ISSUE TITLE
Emulate EXT_direct_state_access to remove extension dependency

### ISSUE THREAD

**Issue by kvark** [baldurk/renderdoc#428]
It doesn't look like Mesa is going to implement the EXT version, while ARB is already supported, for example, on my current machine:
> Mesa DRI Intel(R) HD Graphics 5500 (Broadwell GT2)

If we can't drop EXT, perhaps we could use separate code paths for ARB and EXT, supporting both of them.

**Comment by baldurk** [comment]
It's not practical to use ARB and EXT, there are semantic differences between them and it'd make kind of a mess of the code, plus ARB doesn't have great driver support so it's a dubious benefit.

The solution is to emulate EXT just by doing something like:

```cpp
glGet(BINDING, &prev);
glBind(foo);
glDoThingToFoo();
glBind(prev);
```

which also supports systems without either extension.

**Comment by kvark** [comment]
@jrmuizel suggested a workaround by linking with https://github.com/p3/regal, which seems to support DSA

**Comment by baldurk** [comment]
I think that would be kind of like hammering a nail with a bulldozer. The code doesn't call too many EXT_dsa functions, so it just needs a few dozen emulation functions added since it's easy to do along the pattern I have above. The tricky part is being able to verify that it works and I haven't forgotten one somewhere.

I think if I set up a VM and run mesa's software renderer I should be able to limit it to just 3.2. It's a bit awkward but that's probably the best way for me to test. I'd like to do this right, so at the same time removing other assumptions and being able to replay on 3.2 and above would be ideal - I'd just have some features like e.g. histogram/range autofit turn themselves off if compute support isn't present.

**Comment by baldurk** [comment]
I've pushed up changes that reduce the assumptions RenderDoc makes on replay. Rather than needing 4.3 + `GL_ARB_buffer_storage` + `GL_EXT_direct_state_access`, it now requires a 3.2 context plus at least:

* `GL_ARB_vertex_attrib_binding`
* `GL_ARB_program_interface_query`
* `GL_ARB_shading_language_420pack`
* `GL_ARB_separate_shader_objects`
* `GL_ARB_explicit_attrib_location`
* `GL_ARB_sampler_objects`

Which hopefully should be pretty widely available as none of those require much hardware support (if any). I was testing on mesa 10.6.0 which is fairly old at this point with the softpipe renderer, and I got capture and replay to work fine. Note that earlier versions might not have a couple of those, I went back about as far as I could get away with. For reference, Ubuntu 16.04 LTS and 16.10 both have newer mesa versions than that shipping.

There are obviously caveats - some features require compute shaders or shader image/load store, so they just get turned off and you can't use them. Perhaps in future a non-compute or non-load/store fallback path could be written. Also of course if the captured log uses some feature X then it will be required on replay. Currently there's no checking that features are available on replay, since it's hard to tell exactly which extensions a program actually used.

Let me know how it goes for you with these latest changes.

**Comment by baldurk** [comment]
This should be fixed now in v0.33. If you have any problems with it still let me know, and I'm curious to hear if anyone finds those extension requirements onerous. In theory many of them could be removed, but it would take more work and didn't seem worthwhile.
